rust/blockstore/src/arrow/ordered_blockfile_writer.rs

```diff
@@ -0,0 +1,962 @@
+            }
+            if !removed {
+                self.root
+                    .sparse_index
+                    .set_count(delta.id(), delta.len() as u32)
+                    .map_err(|e| Box::new(e) as Box<dyn ChromaError>)?;
+                let block = self.block_manager.commit::<K, V>(delta);
+                blocks.push(block);
+            }
+        }
+
+        // MIGRATION(10/15/2024 @hammadb) Get all the blocks and manually update the sparse index
+        if self.root.version == Version::V1 {
+            self.root.version = Version::V1_1;
+            let block_ids;
+            // Guard the sparse index data access with a lock
+            // otherwise we have to hold the lock across an await
+            {
+                let sparse_index_data = self.root.sparse_index.data.lock();
+                block_ids = sparse_index_data
+                    .forward
+                    .values()
+                    .filter(|block_id| !handled_blocks.contains(block_id))
+                    .copied()
+                    .collect::<Vec<Uuid>>();
+            }
+            for block_id in block_ids.iter() {
+                let block = self.block_manager.get(block_id).await;
+                match block {
+                    Ok(Some(block)) => {
+                        match self
+                            .root
+                            .sparse_index
+                            .set_count(*block_id, block.len() as u32)
+                        {
+                            Ok(_) => {}
+                            Err(e) => {
+                                return Err(Box::new(e));
+                            }
+                        }
+                    }
+                    Ok(None) => {
+                        return Err(Box::new(ArrowBlockfileError::BlockNotFound));
+                    }
+                    Err(e) => {
+                        return Err(Box::new(e));
+                    }
+                }
+            }
+        }
+
+        let flusher = ArrowBlockfileFlusher::new(
+            self.block_manager,
+            self.root_manager,
+            blocks,
+            self.root,
+            self.id,
+        );
+
+        Ok(flusher)
+    }
+
+    async fn swap_current_delta<K: ArrowWriteableKey, V: ArrowWriteableValue>(
+        &self,
+        new_delta_block_id: &Uuid,
+        new_delta_end_key: Option<CompositeKey>,
+    ) -> Result<(), Box<dyn ChromaError>> {
+        self.flush_current_delta::<K, V>().await?;
+
+        let new_delta = self
+            .block_manager
+            .fork::<K, V, OrderedMergeBlockDelta>(new_delta_block_id)
+            .await
+            .map_err(|e| Box::new(e) as Box<dyn ChromaError>)?;
+
+        self.root
+            .sparse_index
+            .replace_block(*new_delta_block_id, new_delta.id);
+
+        let mut current_delta = self.current_block_delta.lock().await;
+        *current_delta = Some((new_delta, new_delta_end_key));
```

Comment:
decision: custom impl of Clone that panics, get rid of locks

Reply:
I decided not to go down this path for now as it requires changing the signature of `.set()`/`.delete()` to take `&mut self` which touches _everything_ above it

Reply:
> Is there a downside to pushing the onus of synchronization up to the caller

There are tremendous downsides. We want the Blockfile to be a threadsafe abstraction. Pushing complexity up makes all other code incredibly complicated. The blockfile is a concurrency safe key-value store.

Reply:
I made this comment a while ago haha. I agree that in general the blockfile writer can perform more granular and correct locking than users. However, given that ordered writing mode is currently inherently single-threaded, requiring users to `Arc<Mutex<>>` it if they want to share it across threads and otherwise not requiring a lock didn't seem that far out there. I don't think the semantic split between the two writing modes is worth it though.